File: node.js/백준/Gold/3109. 빵집/빵집.js
// 3109 빵집
const fs = require("fs");

const settings = () => {
    // const input = fs.readFileSync("input.txt").toString().trim().split("\n");
    const input = fs.readFileSync("/dev/stdin").toString().trim().split("\n");
    const [R, C] = input[0].split(" ").map(Number);
    const ground = [];
    input.slice(1).forEach((e) => ground.push(e));
    return [R, C, ground];
}


const main = (R, C, ground) => {
    const makePipeLine = (row) => {
        const stack = [[row, 0]];
        
        while (stack.length > 0) {
            const [i, j] = stack.pop();
            visited[i][j] = 1
            if ( j === C-1) return true;

            for (let di = 1; di > -2; di --){
                const [ni,nj] = [di+i, j+1];
                if (0<=ni && ni <R && nj < C && ground[ni][nj] !== "x" && visited[ni][nj] !== 1){
                    stack.push([ni,nj]);
                }
            }

        }   
        return false;
    }
    let cnt = 0;
    const visited = [];
    for (let r=0; r<R;r++){
        visited.push(Array(C).fill(0));
    }


    for (let i=0; i<R; i++){
        if (ground[i][0] !== "x"){
            cnt += makePipeLine(i);
        }
    }
    console.log(cnt);
}

const [R, C, ground] = settings();

main(R, C, ground);

File: Python/백준/Gold/3109. 빵집/빵집.py
# 3109 빵집
import sys
input = sys.stdin.readline

R, C = map(int, input().split())
ground = tuple(input().strip() for _ in range(R))


def makePipeLine(row):
    stack = [(row,0)] # i, j, cnt

    while stack:
        i, j = stack.pop()
        visited[i][j] = 1
        if j == C-1:
            return 1
        

        for di in range(1, -2, -1):
            ni,nj = i+di, j+1
            if 0 <=ni < R and nj < C:
                if ground[ni][nj] != 'x' and not visited[ni][nj]:
                    stack.append((ni,nj))
    
    return 0

ans = 0
visited = [[0] * C for _ in range(R)]

for i in range(R):
    maxv = 0
    if ground[i][0] != "x":
        visited[i][0] = 1
        ans += makePipeLine(i)
# [print(visited[k]) for k in range(R)]
print(ans)  